src/main/java/gregtech/api/unification/material/MaterialBuilder.java

```diff
@@ -0,0 +1,478 @@
+package gregtech.api.unification.material;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import gregtech.api.fluids.fluidType.FluidType;
+import gregtech.api.unification.Element;
+import gregtech.api.unification.material.Material.MaterialInfo;
+import gregtech.api.unification.material.info.MaterialFlag;
+import gregtech.api.unification.material.info.MaterialFlags;
+import gregtech.api.unification.material.info.MaterialIconSet;
+import gregtech.api.unification.material.properties.*;
+import gregtech.api.unification.stack.MaterialStack;
+import net.minecraft.enchantment.Enchantment;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+
+public abstract class MaterialBuilder implements IMaterialBuilder {
+
+    protected abstract <T extends IMaterialProperty<T>> void setProperty(PropertyKey<T> key, T property);
```

Comment:
As this is protected, there wouldn't be any great ways for addons to create their own material properties other than just extending this class itself, which might cause quite a few conflicts down the line. Perhaps this could be made public (and then put into the interface)?